WTF: Haiku filesystem and language support

pulkomandy opened the pull request:
#### 03df07d307c1b20968adbd66b681ada56da2c10a
<pre>
WTF: Haiku filesystem and language support
<a href="https://bugs.webkit.org/show_bug.cgi?id=305764">https://bugs.webkit.org/show_bug.cgi?id=305764</a>

Reviewed by Adrian Perez de Castro.

Upstream changes to filesystem and language access for Haiku: finding
the tmp directory, user configuration directory, and user preferred
languages using native APIs where no POSIX standard way is applicable.

No new tests: no changes to behavior of exisitng platforms.

* Source/WTF/wtf/PlatformHaiku.cmake: Add new files to the build.
* Source/WTF/wtf/haiku/FileSystemHaiku.cpp: Added.
Use native APIs to determine user settings directory.
* Source/WTF/wtf/haiku/LanguageHaiku.cpp: Added.
Use native APIs to detect user language settings.
* Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
(WTF::FileSystemImpl::fileCreationTime): enable shared code for Haiku
(WTF::FileSystemImpl::temporaryFileDirectory): use find_directory

Canonical link: <a href="https://commits.webkit.org/306080@main">https://commits.webkit.org/306080@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7a92255b9c2d02cc20adbc7af0f631bfd59fd014

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/140307 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/12688 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/1818 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/148595 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/93385 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/13400 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/12842 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/107520 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/93385 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/143257 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/10315 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/125607 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/88414 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/9936 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/7483 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/8740 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/132280 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/119169 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/167/builds/1620 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/151251 "Built successfully") | 
| [✅ 🛠 🧪 jsc-debug-arm64](https://ews-build.webkit.org/#/builders/171/builds/1103 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/12376 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/1687 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/115847 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/12388 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/10595 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/116183 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/29514 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/11267 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/122093 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/67389 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/12417 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/1547 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/171580 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/12158 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/76115 "Built successfully") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/44524 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/12352 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/12202 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WTF/wtf/haiku/LanguageHaiku.cpp:
```diff
@@ -0,0 +1,65 @@
+static String platformLanguage()
+{
+    static BString locale;
+    static bool initialized = false;
```
Could this be called from different threads? If yes, or unsure, I would use `std::once_flag` + `std::call_once` to avoid race conditions.

If this will only be used from a single thread, I would still use an anonymous lambda to avoid the need for the boolean flag. Something like:

```cpp
static const BString locale = ([]() {
    BLanguage language;
    if (BLocale::Default()->GetLanguage(&language) == B_OK) {
        auto localeId = language.ID();
        localeId.replaceAll('_', '-');
        return localeId;
    }
    return "c";
})();
return String::fromUTF8(locale.String());
```

pulkomandy replied:
The code was reworked and this should be called only from the main thread.

The intitialization function is called from WebKit and WebKitLegacy layers so that will be in place when those are upstreamed.

aperezdc commented on Source/WTF/wtf/haiku/LanguageHaiku.cpp:
```diff
@@ -0,0 +1,65 @@
+{
+    Vector<String> userPreferredLanguages;
+    userPreferredLanguages.append(platformLanguage());
+    return userPreferredLanguages;
```
You could use an initializer list to simplify this:

 
```suggestion
    return { platformLanguage() };
```

(There's nothing wrong with the code, though! Feel free to leave it as-is.)

pulkomandy replied:
Updated the code so it actually supports adding multiple languages now (if the user did set that up in their Locale preferences).

aperezdc commented on Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
```diff
@@ -158,6 +163,16 @@ static const char* temporaryFileDirectory()
+    static char buffer[B_PATH_NAME_LENGTH];
+    if (buffer[0] == '\0') {
+        BPath path;
+        if (find_directory(B_SYSTEM_TEMP_DIRECTORY, &path) == B_OK)
```
Does this really work? Looking at the [documentation](https://www.haiku-os.org/docs/api/FindDirectory_8h.html) I don't see a version of this function that takes two arguments, and it is not [in the FindDirectory.h header](https://github.com/haiku/haiku/blob/master/headers/os/storage/FindDirectory.h).

pulkomandy replied:
It's at the very end of the header:

https://github.com/haiku/haiku/blob/master/headers/os/storage/FindDirectory.h#L207

This version takes 4 parameters, but the last two are optional with default values.

aperezdc replied:
Okay, I was clearly not caffeinated enough when I went through the docs and header 🤦🏼

aperezdc replied:
Also, I was checking a bit and I think this may end up being called from different threads so the initialization is racy as it is.

aperezdc reviewed (approved):
@pulkomandy This is looking definitely nicer now, thanks! (and r=me)

pulkomandy committed:
WTF: Haiku filesystem and language support
https://bugs.webkit.org/show_bug.cgi?id=305764

Reviewed by Adrian Perez de Castro.

Upstream changes to filesystem and language access for Haiku: finding
the tmp directory, user configuration directory, and user preferred
languages using native APIs where no POSIX standard way is applicable.

No new tests: no changes to behavior of exisitng platforms.

* Source/WTF/wtf/PlatformHaiku.cmake: Add new files to the build.
* Source/WTF/wtf/haiku/FileSystemHaiku.cpp: Added.
Use native APIs to determine user settings directory.
* Source/WTF/wtf/haiku/LanguageHaiku.cpp: Added.
Use native APIs to detect user language settings.
* Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
(WTF::FileSystemImpl::fileCreationTime): enable shared code for Haiku
(WTF::FileSystemImpl::temporaryFileDirectory): use find_directory

Canonical link: https://commits.webkit.org/306080@main
